src/haz3lcore/statics/Coverage.re

```diff
@@ -0,0 +1,504 @@
+  let false_ = Ap("false", None);
+};
+
+module Ctr = {
+  module M = {
+    // Ctrs are Constructors equipped with arities.
+    [@deriving (show({with_path: false}), sexp, yojson)]
+    type t = {
+      ctr: Constructor.t,
+      num_args: int,
+      string_key: string,
+    };
+    let compare =
+        (
+          {ctr: _, num_args: _, string_key: string_key1},
+          {ctr: _, num_args: _, string_key: string_key2},
+        ) =>
+      compare(string_key1, string_key2);
+  };
+  include M;
+  module Map = MapUtil.Make(M);
+  module Set = Set.Make(M);
+
+  let mk = (ctr, num_args) => {
+    let string_key = string_of_int(num_args) ++ "~" ++ ctr;
+    {ctr, num_args, string_key};
+  };
+  let num_args_of = (ctr: t): int => ctr.num_args;
+
+  // Ctrs for primitive types
+  // Names here should not be anotherwise valid sum type constructor names.
+  let of_int = n => mk(string_of_int(n), 0);
+  let of_float = x => mk(string_of_float(x), 0);
+  let of_string = s => mk("\"" ++ s, 0); // don't need closing quote, just need to distinguish from others
+  let tuple_ctr: int => t = n => mk("tuple", n);
+  let nil_ctr = mk("nil", 0);
+  let cons_ctr = mk("cons", 1);
+  let true_ctr = mk("true", 0);
+  let false_ctr = mk("false", 0);
+
+  // used when not all constructors have been seen to handle the unseen cases when a wildcard appears
+  let default_ctr = mk("_", 0);
+
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type arity = list(Typ.t);
+
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type all_ctrs =
+    | Unknown
+    | Infinite
+    | Finite(Map.t(arity));
+
+  let arity_of = (ctr, all_ctrs: all_ctrs): arity =>
+    switch (all_ctrs) {
+    | Unknown => List.init(ctr.num_args, _ => Typ.Fresh.tunknown(Internal))
+    | Infinite => List.init(ctr.num_args, _ => Typ.Fresh.tunknown(Internal))
+    | Finite(all_ctrs) =>
+      switch (Map.find_opt(ctr, all_ctrs)) {
+      | Some(arity) => arity
+      | None => List.init(ctr.num_args, _ => Typ.Fresh.tunknown(Internal))
+      }
+    };
+
+  let rec all_ctrs_of_typ = (ty: Typ.t): all_ctrs =>
+    switch (ty.term) {
+    | Sum(map) =>
+      Finite(
+        map
+        |> List.filter_map(
+             fun
+             | ConstructorMap.Variant(ctr, _, None) =>
+               Some((mk(ctr, 0), []))
+             | Variant(ctr, _, Some(arg_ty)) =>
+               Some((mk(ctr, 1), [arg_ty]))
+             | BadEntry(_) => None,
+           )
+        |> Map.of_list,
+      )
+    | Rec(_) => all_ctrs_of_typ(Typ.unroll(ty))
+    | Prod(elts) =>
+      Finite(Map.singleton(tuple_ctr(List.length(elts)), elts))
+    | TupLabel(_, ty) => Finite(Map.singleton(tuple_ctr(1), [ty]))
+    | List(elt_ty) =>
+      Finite(
+        Map.of_list([
+          (nil_ctr, []),
+          (cons_ctr, [Typ.Fresh.tprod([elt_ty, ty])]),
+        ]),
+      )
+    | Bool => Finite(Map.of_list([(true_ctr, []), (false_ctr, [])]))
+    | Unknown(_) => Unknown
+    | Int
```

Comment:
looks like this is going reject a case that actually handles all integers as non-exhaustive? literally unusable

Reply:
ha yes, I'll leave a comment noting this serious correctness issue